### DISCUSSION TITLE
getting all installed packages in composer plugin events

### DISCUSSION THREAD

**Discussion by doppynl** [composer/composer#10506] [Q&A / Support] [answered]
I'm writing a composer plugin and have a need for a list of all packages during a `pre-autoload-dump` or `post-install-cmd` event. With a Package object as I need some information from the `extra` config of the package.

So far I've only managed to do this to the extend of keeping track of updated/installed packages via the `post-package-update/install` events. But this only works for packages that change, not for packages that are already installed and don't change.

I can manage it manually by doing a `composer reinstall`, but that would require running a second command.

The goal of getting all the packages is so I can set some symlinks for certain packages that have some `extra` configuration for them. Please don't ask why, it has to do with legacy things.

Basically, if I can do one of the following 2 things in `pre-autoload-dump` en `post-install-cmd` events I would be really happy:

- get a complete list of all installed packages (as `Package` objects)?

OR

- both these things
  - get a list of all installed packages (as `Package` objects) by a specifc package-type
  - get a package (as `Package` object) by its name

**Comment by Seldaek** [comment, marked as answer]
On the event object you have `->getComposer()->getRepositoryManager()->getLocalRepository()` which is a package repo, you can call getCanonicalPackages() on it, or findPackage(), that should do what you need.

**Reply by doppynl** [reply]
I thought I did a thorough search on a method like that, this one I apparently missed.

The method is exactly what I need. Thanks for the help and quick answer!
